modules/core/include/opencv2/core/dualquaternion.inl.hpp

```diff
@@ -0,0 +1,392 @@
+inline DualQuat<T>& DualQuat<T>::operator+=(const DualQuat<T> &q)
+{
+    *this = *this + q;
+    return *this;
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::operator*(const DualQuat<T> &q) const
+{
+    Quat<T> A = getRealQuat();
+    Quat<T> B = getDualQuat();
+    Quat<T> C = q.getRealQuat();
+    Quat<T> D = q.getDualQuat();
+    return DualQuat<T>::createFromQuat(A * C, A * D + B * C);
+}
+
+template <typename T>
+inline DualQuat<T>& DualQuat<T>::operator*=(const DualQuat<T> &q)
+{
+    *this = *this * q;
+    return *this;
+}
+
+template <typename T>
+inline DualQuat<T> operator+(const T a, const DualQuat<T> &q)
+{
+    return DualQuat<T>(a + q.w, q.x, q.y, q.z, q.w_, q.x_, q.y_, q.z_);
+}
+
+template <typename T>
+inline DualQuat<T> operator+(const DualQuat<T> &q, const T a)
+{
+    return DualQuat<T>(a + q.w, q.x, q.y, q.z, q.w_, q.x_, q.y_, q.z_);
+}
+
+template <typename T>
+inline DualQuat<T> operator-(const DualQuat<T> &q, const T a)
+{
+    return DualQuat<T>(q.w - a, q.x, q.y, q.z, q.w_, q.x_, q.y_, q.z_);
+}
+
+template <typename T>
+inline DualQuat<T>& DualQuat<T>::operator-=(const DualQuat<T> &q)
+{
+    *this = *this - q;
+    return *this;
+}
+
+template <typename T>
+inline DualQuat<T> operator-(const T a, const DualQuat<T> &q)
+{
+    return DualQuat<T>(a - q.w, -q.x, -q.y, -q.z, -q.w_, -q.x_, -q.y_, -q.z_);
+}
+
+template <typename T>
+inline DualQuat<T> operator*(const T a, const DualQuat<T> &q)
+{
+    return DualQuat<T>(q.w * a, q.x * a, q.y * a, q.z * a, q.w_ * a, q.x_ * a, q.y_ * a, q.z_ * a);
+}
+
+template <typename T>
+inline DualQuat<T> operator*(const DualQuat<T> &q, const T a)
+{
+    return DualQuat<T>(q.w * a, q.x * a, q.y * a, q.z * a, q.w_ * a, q.x_ * a, q.y_ * a, q.z_ * a);
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::operator/(const T a) const
+{
+    return DualQuat<T>(w / a, x / a, y / a, z / a, w_ / a, x_ / a, y_ / a, z_ / a);
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::operator/(const DualQuat<T> &q) const
+{
+    return *this * q.inv();
+}
+
+template <typename T>
+inline DualQuat<T>& DualQuat<T>::operator/=(const DualQuat<T> &q)
+{
+    *this = *this / q;
+    return *this;
+}
+
+template <typename T>
+std::ostream & operator<<(std::ostream &os, const DualQuat<T> &q)
+{
+    os << "DualQuat " << Vec<T, 8>{q.w, q.x, q.y, q.z, q.w_, q.x_, q.y_, q.z_};
+    return os;
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::exp() const
+{
+    DualQuat<T> v(0, x, y, z, 0, x_, y_, z_);
+    DualQuat<T> normV = v.norm();
+    DualQuat<T> sin_normV(std::sin(normV.w), 0, 0, 0, normV.w_ * std::cos(normV.w), 0, 0, 0);
+    DualQuat<T> cos_normV(std::cos(normV.w), 0, 0, 0, -normV.w_ * std::sin(normV.w), 0, 0, 0);
+    DualQuat<T> exp_w(std::exp(w), 0, 0, 0, w_ * std::exp(w), 0, 0, 0);
+    if (normV.w < CV_DUAL_QUAT_EPS)
+    {
+        return exp_w * (cos_normV + v);
+    }
+    DualQuat<T> k = sin_normV / normV;
+    return exp_w * (cos_normV + v * k);
+}
```

Comment:
Now I think to use Taylor expansion is better.
`exp(p+eq) = exp(p)+e qexp(p)`

Reply:
Be careful with that; a derivative for non-commutative functions can be tricky.
I use the following expression: `exp(w_real + e*w_dual) = exp(w_real) + e*J_exp_quat(w_real)*w_dual`
where `J_exp_quat()` is a 4x4 jacobian of quaternionic exp(w_real), `w_dual` is a vector.
```
J_exp_quat(w+ijk*v) = exp(w)*(dw*exp(ijk*v) + d(exp(v)))
d(exp(v)) = -sinc(nv)* v^T*dv + ijk*csiii(nv) *(v*v^t)*dv + ijk*sinc(nv)*dv
nv = norm(v)
sinc(x) = sin(x)/x
csiii(x) = (cos(x) - sin(x)/x)/x/x
```
so the resulting matrix is:
```
[ cos(nw)    ; -sinc(nv) * v^T
  sinc(nv)*v ; csiii(nv) * (v*v^T) + I_3*sinc(nv) ] * exp(w)
```
I guess the logarithm can be calculated similarly: first we get log(w_real), then we use inverse of J_exp_quat for dual part.

Reply:
I'm sorry I'm a bit confused with this, but I find a [paper](https://hal.archives-ouvertes.fr/hal-01122406v5/document) that has a detailed description of this. I will change it in the next commit.

Reply:
I've read some parts of the paper you've provided. The jacobians they provide are more about functions to or from SO(3), this is why they replace + and - by those Exp- and Log- combined operators.
What we need here is a derivative from one linear space (exp argument) to another one (resulting quaternion). What I wrote in the comment above is exactly that kind of jacobian. I got it with usual derivation rules like chain rule, etc.

<details>
  <summary>Here's how I've got the formulas above, you can check it</summary>

```
exp(v) = exp(1/2*alpha*n) =
< alpha := 2*norm(v), n := v/norm(v) > =
cos(alpha/2) + ijk*sin(alpha/2)*n

d(exp(v)) = d(cos(alpha/2) + ijk*sin(alpha/2)*n) =
d(cos(norm(v)) + ijk*sin(norm(v))/norm(v)*v) =
< sin(x)/x = sinc(x) > =
d(cos(norm(v)) + ijk*sinc(norm(v))*v) =
-sin(norm(v))*d(norm(v)) + ijk*d(sinc(norm(v)))*v + ijk*sinc(norm(v))*dv = (*)

norm(v) = sqrt(dot(v, v))
d(sqrt(x)) = 1/2/sqrt(x)*dx
d(dot(x, y)) = dot(dx, y) + dot(x, dy)
d(norm(v)) = d(sqrt(dot(v, v))) = 1/2/sqrt(dot(v,v))*d(dot(v, v)) = 1/2/sqrt(dot(v, v))*(2*dot(dv, v)) =
dot(v, dv)/norm(v) = v^T/norm(v)*dv
d(sinc(x)) = d(sin(x)/x) = (dsin(x)*x - sin(x)*dx)/x/x = (cos(x)*x - sin(x))/x/x*dx =
(cos(x) - sin(x)/x)/x*dx
d(sinc(norm(v)) = dsinc {x=norm(v)} * d(norm(v)) =
(cos(x) - sinc(x))/x*d(x) =
(cos(norm(v)) - sinc(norm(v)))/norm(v)*d(norm(v)) =
(cos(norm(v)) - sinc(norm(v)))/norm(v)*dot(v, dv)/norm(v) =
(cos(norm(v)) - sinc(norm(v)))/norm2(v)*dot(v, dv) =
< (cos(x) - sinc(x))/x/x = csiii(x) >
csiii(norm(v)) * dot(v, dv)

(*)
= -sin(norm(v))*dot(v, dv)/norm(v) + ijk*csiii(norm(v))*dot(v, dv)*v + ijk*sinc(norm(v))*dv =
= -sinc(norm(v))*dot(v, dv) + ijk*csiii(norm(v))*v*dot(v, dv) + ijk*sinc(norm(v))*dv =
= -sinc(norm(v))*v^T*dv + ijk*csiii(norm(v))*(v*v^T)*dv + ijk*sinc(norm(v))*I_3*dv =
= (-sinc(norm(v))*v^T + ijk*csiii(norm(v))*(v*v^T) + ijk*sinc(norm(v))*I_3)*dv
```

</details>

The only tricky place here is `d(...)` which can be treated as `d/dt (...)` where `t` is some extra variable. I used [this](https://core.ac.uk/download/pdf/33363595.pdf) as a hint how to obtain the formulas.
I've also double-checked them by calculating the exponential form of dual quaternion. I've got exactly the same jacobian in dual part.

Reply:
Thank you very much for giving such a detailed explanation of the calculation for d(exp). I have written the calculation method of exp and log you've above into the code, and found no problem.